### ISSUE TITLE
KCM keeps polling half-closed sockets

### ISSUE THREAD

**Issue by fihuer** [heimdal/heimdal#808]
**Describe the bug**
KCM daemon busy looping on half-closed sockets

**To Reproduce**
Not able to reproduce it (yet)

**Expected behavior**
Half-closed sockets should be closed, KCM should not poll on them, keeping a whole CPU busy.

**Machine:**
 - OS: RHEL 8.3, Heimdal 7.7.0

**Additional context**
[This line's](https://github.com/heimdal/heimdal/blob/a56c3ad8199c418a3e3458d3f433608043b65bed/lib/ipc/server.c#L688) conditions seems suspect to me. GDB'ing the KCM process show that half-closed sockets may have a WAITING_READ and WAITING_CLOSE flags, and this line skips the `close` call

**Comment by jaltman** [comment]
In your build is HAVE_GCD defined?    
If so, does it make a difference if 
`    if (c->flags & (WAITING_READ|WAITING_WRITE))
	return 0;`
is moved after the HAVE_GCD block?

**Comment by fihuer** [comment]
No, we don't have this flag defined (RedHat systems). So yes it would make a difference for us.

**Comment by lhoward** [comment]
I don't really understand the code enough to know if that's the right solution. @fihuer, have you tried moving the `HAVE_GCD` guard in `maybe_close()` to see if it fixes your problem?

**Comment by nicowilliams** [comment]
@lhoward does Darwin have a better kcm than Heimdal?

**Comment by lhoward** [comment]
@nicowilliams it does have a lot of changes but many are Apple and NTLM specific. Since Big Sur, they no longer use it and I'm not sure if the replacement is open source.

**Comment by fihuer** [comment]
@nicowilliams Is this issue supposed to be fixed with f3517c0 ?
